packages/@expo/cli/src/utils/telemetry/index.ts

```diff
@@ -0,0 +1,28 @@
+import { DetachedClient } from './DetachedClient';
+import { RudderClient } from './RudderClient';
+import type { TelemetryClient, TelemetryEvent, TelemetryProperties } from './types';
+import { env } from '../env';
+
+const debug = require('debug')('expo:telemetry') as typeof console.log;
+
+function getClient(): TelemetryClient | null {
+  if (env.EXPO_NO_TELEMETRY) return null;
+
+  const client = env.EXPO_NO_TELEMETRY_DETACH
+    ? new RudderClient() // Block the CLI process when sending telemetry, useful for testing
+    : new DetachedClient(); // Do not block the CLI process when sending telemetry
+
+  process.once('SIGINT', () => client.flush());
+  process.once('SIGTERM', () => client.flush());
+  process.once('beforeExit', () => client.flush());
+
+  return client;
+}
+
+/** The singleton telemetry instance */
+export const telemetry = getClient();
+
+export async function logEventAsync(event: TelemetryEvent, properties?: TelemetryProperties) {
+  await telemetry?.record(event, properties);
+  debug('Telemetry event recorded:', event);
```

Comment:
Maybe move this `debug` statement to the telemetry classes itself. That way, we can deprecate this method in favor of just:

```js
import { telemetry } from './utils/telemetry';

telemetry?.record(...);
```